### PR TITLE
Feature Request: Implement various sorting algorithms

### LINKED ISSUE #660: Feature Request: Implement various sorting algorithms
#### Description of the problem
We need to enhance the project by implementing a collection of commonly used sorting algorithms. Adding these algorithms will improve functionality and provide users with efficient tools for sorting data. The goal is to implement a variety of sorting algorithms, each well-documented and optimized for performance. Below is a list of sorting algorithms to be added:

- Shell Sort
- Radix Sort
- Tim Sort
- Tree Sort
- Cycle Sort
- Odd-even Merge Sort
- Comb Sort
- Bitonic Sort

#### Example of the problem
This is a feature request rather than a bug, so no specific buggy code exists to reproduce. However, as an example of what’s needed, here’s a placeholder for where sorting functionality could be added:

```python
# Example of missing sorting functionality
def sort_data(data):
    # Placeholder - no sorting implemented yet
    return data

# Desired usage
data = [64, 34, 25, 12, 22, 11, 90]
sorted_data = sort_data(data)
print(sorted_data)  # Expected: [11, 12, 22, 25, 34, 64, 90]

### PR DESCRIPTION
**Pull request by arvinder004**
Fixes #660  

#### Brief description of what is fixed or changed
This pull request introduces the Shell Sort algorithm to the sorting module, implemented in a style consistent with existing algorithms like `insertion_sort`. It includes:
- A `shell_sort` function in `algorithms.py` that supports partial sorting (`start` to `end`), custom comparators (`comp`), and backend switching (`backend`).
- Comprehensive tests in `test_algorithms.py` using `pytest` and `pytest-cov`, covering typical cases, edge cases (empty, single-element), sorted/reverse-sorted lists, duplicates, and negative numbers.
- Documentation with examples in the function's docstring, matching the project's format.

The implementation uses a simple gap sequence (`n // 2`) for clarity, with potential for future optimization using advanced sequences

### CODE REVIEW

**Comment by arvinder004** [comment]
#### References to other Issues or PRs or Relevant literature
Fixes #660 
#### Brief description of what is fixed or changed
This pull request introduces the Radix Sort algorithm to the sorting module, implemented in a style consistent with existing algorithms like `insertion_sort` and `shell_sort`. It includes:
- A `radix_sort` function in `sorting.py` that supports partial sorting (`start` to `end`), custom comparators (`comp`, ignored as it’s non-comparison-based), and backend switching (`backend`). The implementation sorts non-negative integers using a counting sort subroutine for each digit.
- Comprehensive tests in `test_sorting.py` using `pytest` and `pytest-cov`, covering typical cases (multi-digit numbers), edge cases (empty, single-element), sorted/reverse-sorted lists, duplicates, large numbers, and error handling for negative numbers.
- Documentation with examples in the function's docstring, matching the project's format.

**Comment by arvinder004** [comment]
Sure sir, I got it. Thankyou for taking out your valueable time to review my PR and suggesting changes

**Comment by arvinder004** [comment]
@TripleCamellya Please check my latest commit, I have done all the suggested changes. I needed a seperated test file so kept the previous one for that

**Comment by TripleCamellya** [comment]
> ## [Codecov](https://app.codecov.io/gh/codezonediitj/pydatastructs/pull/661?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=codezonediitj) Report
> Attention: Patch coverage is `92.15686%` with `4 lines` in your changes missing coverage. Please review.
> 
> > Project coverage is 97.278%. Comparing base [(`f4c1677`)](https://app.codecov.io/gh/codezonediitj/pydatastructs/commit/f4c1677965d7385c11e134b7c0e6123ae8854a98?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=codezonediitj) to head [(`99cd580`)](https://app.codecov.io/gh/codezonediitj/pydatastructs/commit/99cd58071649db2c57b8f5820c3667f4d9d7bbe3?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=codezonediitj).
> > Report is 3 commits behind head on main.
> 
> [Files with missing lines](https://app.codecov.io/gh/codezonediitj/pydatastructs/pull/661?dropdown=coverage&src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=codezonediitj)	Patch %	Lines
> [pydatastructs/linear_data_structures/algorithms.py](https://app.codecov.io/gh/codezonediitj/pydatastructs/pull/661?src=pr&el=tree&filepath=pydatastructs%2Flinear_data_structures%2Falgorithms.py&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=codezonediitj#diff-cHlkYXRhc3RydWN0cy9saW5lYXJfZGF0YV9zdHJ1Y3R1cmVzL2FsZ29yaXRobXMucHk=)	92.156%	[4 Missing ⚠️ ](https://app.codecov.io/gh/codezonediitj/pydatastructs/pull/661?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=codezonediitj)
> Additional details and impacted files
> 🚀 New features to boost your workflow:

1. There are unreachable branches in your algorithm. Line 1981 said start must less than end, then for line 1984-1986, n must more than 0. So the judge is unreachable.
2. Your test code didn't use DynamicArray, so the branch for DynamicArray failed. You can refer to _test_common_sort().

**Comment by arvinder004** [comment]
@TripleCamellya Sir please see this PR

**Comment by TripleCamellya** [comment]
Just wait for czgdp1807's review.

**Comment by arvinder004** [comment]
okay sir.
@czgdp1807 sir please review.

**Review by czgdp1807 — APPROVED** [review]

File: pydatastructs/linear_data_structures/algorithms.py [hunk 1]
```diff
@@ -1850,3 +1852,164 @@ def partition(array, lower, upper):
...
+        is to be sorted.
+        Optional, by default the index
+        of the last position filled.
+    comp: lambda/function
+        The comparator which is to be used
+        for sorting. If the function returns
+        False then only swapping is performed.
+        Optional, by default, less than or
+        equal to is used for comparing two
+        values.
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:1954]:
```suggestion
```

**Review by czgdp1807 — APPROVED** [review]
